### PR TITLE
Feat 568 triple chevron

### PR DESCRIPTION
**Pull request by ZelboK**
This ports `triple_chevron_launch` from thrust to CUB. As things stand it makes more sense to decouple the two. However, this PR does not touch the thrust repo. I can do that in this PR or in a follow up PR if it is desired. I thought it would be better to keep the PR smaller. 

Currently all tests pass on my 3080 sm86 except for one, `test_allocator.cu` for `util_allocator.cuh`. However it is important to note that this happens on main as well, so it is likely irrelevant to my changes. If it matters, only line 130 of `test_allocator.cu` assertion fails. If that one is commented out, all tests pass. 

Please pay attention to the differences in macros in `triple_chevron_launch.cuh` - I tried to follow what seemed to be idiomatic for CUB.

### CODE REVIEW

File: .devcontainer/cuda12.2-gcc12/devcontainer.json [hunk 1]
```diff
@@ -7,12 +7,10 @@
...
     "/bin/bash",
     "-c",
-    "mkdir -m 0755 -p ${localWorkspaceFolder}/.{aws,cache,config}"
+    "mkdir -m 0755 -p ${localWorkspaceFolder}/.{cache,config,sccache}"
   ],
   "containerEnv": {
-    "SCCACHE_REGION": "us-east-2",
-    "SCCACHE_BUCKET": "rapids-sccache-devs",
-    "VAULT_HOST": "https://vault.ops.k8s.rapids.ai",
+    "SCCACHE_DIR": "${containerWorkspaceFolder}/.sccache",
```
Comment (ZelboK) [inline .devcontainer/cuda12.2-gcc12/devcontainer.json:13]: I left this, in the PR, not because I think it should be merged(obviously) but rather to highlight how sccache can be used locally for contributors like myself. I am unable to access the cloud o bviously since I'm not an NVIDIA employee. 

Maybe you can comment? @jrhemstad

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  struct _LIBCUDACXX_HIDDEN triple_chevron
+  {
+    typedef size_t Size;
+    dim3 const grid;
+    dim3 const block;
+    Size const shared_mem;
+    cudaStream_t const stream;
+
+    CUB_RUNTIME_FUNCTION // Should this be CUB_RUNTIME_FUNCTION now?
```
Comment (ZelboK) [inline cub/cub/detail/triple_chevron_launch.cuh:51]: This was `THRUST_RUNTIME_FUNCTION` before.

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      k<<<grid, block, shared_mem, stream>>>(args...);
+      return cudaPeekAtLastError();
+    }
+
+    template<class T>
+    size_t __device__
+    align_up(size_t offset) const
+    {
+      size_t alignment = THRUST_NS_QUALIFIER::cuda_cub::alignment_of<T>::value;
+      return alignment * ((offset + (alignment - 1))/ alignment);
```
Comment (ZelboK) [inline cub/cub/detail/triple_chevron_launch.cuh:74]: Is this way of accessing the namespace ok? Since `alignment_of` exists in thrust(and is seemingly not used at all through out CUB), does it make sense to continue relying on it? I was curious to know if it would be better to decouple the two. Thrust makes use of `alignment_of` more. Just wondering if it's best to decouple as much as possible from Thrust for CUB.

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    __device__
+    void fill_arguments(char* buffer,
+                     size_t offset,
+                     Arg const& arg,
+                     Args const& ... args) const
+    {
+      fill_arguments(buffer, copy_arg(buffer, offset, arg), args...);
+    }
+
+    #ifdef THRUST_RDC_ENABLED
```
Comment (ZelboK) [inline cub/cub/detail/triple_chevron_launch.cuh:109]: I am not sure if this conditional is needed. To be quite frank I decided it was better to ask rather than go down thsi rabbit hole 😅

File: cub/cub/detail/triple_chevron_launch.cuh (hunk 2 above)
Comment (miscco) [reply cub/cub/detail/triple_chevron_launch.cuh:51]: Lets go with small steps to keep the PR reviewable

File: .devcontainer/cuda12.2-gcc12/devcontainer.json (hunk 1 above)
Comment (miscco) [reply .devcontainer/cuda12.2-gcc12/devcontainer.json:13]: sccache works the same as with the cloud, it just stores aa local cache. So you need to populate it once but thats it

File: cub/cub/detail/triple_chevron_launch.cuh (hunk 3 above)
Comment (miscco) [reply cub/cub/detail/triple_chevron_launch.cuh:74]: That way is indeed necessary to avoid any ambiguity. Note that the macro has a leading :: in the namespace which ensures that we do not catch relative namespaces

File: cub/cub/detail/triple_chevron_launch.cuh (hunk 4 above)
Comment (miscco) [reply cub/cub/detail/triple_chevron_launch.cuh:109]: This is fully needed, please leave as is. 

As said before, try to keep the PRs small and focused on one topic

File: cub/cub/detail/triple_chevron_launch.cuh (hunk 2 above)
Comment (ZelboK) [reply cub/cub/detail/triple_chevron_launch.cuh:51]: My apologies. I felt like since it's being moved to CUB that I should be updating it to use CUB macros rather than thrust to be consistent with the codebase and decouple the two.

File: .devcontainer/cuda12.2-gcc12/devcontainer.json (hunk 1 above)
Comment (ZelboK) [reply .devcontainer/cuda12.2-gcc12/devcontainer.json:13]: Per our discussion, here is the issue https://github.com/NVIDIA/cccl/issues/593

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+#include <cassert>
+
+CUB_NAMESPACE_BEGIN
+
+
+namespace detail
+{
+
+  struct _LIBCUDACXX_HIDDEN triple_chevron
```
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:43]:
```suggestion
  struct _CCCL_ATTRIBUTE_HIDDEN triple_chevron
```

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+ ******************************************************************************/
+#pragma once
+
+#include <thrust/detail/config.h>
+#include <thrust/system/cuda/detail/core/alignment.h>
+#include <thrust/system/cuda/detail/guarded_cuda_runtime_api.h>
+
+#include <cuda/std/version> // _LIBCUDACXX_HIDDEN
+
+#include <cassert>
```
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:35]:
```suggestion
#include <cuda/__cccl_config> // _CCCL_ATTRIBUTE_HIDDEN

#if defined(_CCCL_COMPILER_NVHPC) && defined(_CCCL_USE_IMPLICIT_SYSTEM_DEADER)
#pragma GCC system_header
#else // ^^^ _CCCL_COMPILER_NVHPC ^^^ / vvv !_CCCL_COMPILER_NVHPC vvv
_CCCL_IMPLICIT_SYSTEM_HEADER
#endif // !_CCCL_COMPILER_NVHPC

#include <cub/config.cuh>
#include <cuda/std/type_traits>
#include <nv/target>
```

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    {
+      k<<<grid, block, shared_mem, stream>>>(args...);
+      return cudaPeekAtLastError();
+    }
+
+    template<class T>
+    size_t __device__
+    align_up(size_t offset) const
+    {
+      size_t alignment = THRUST_NS_QUALIFIER::cuda_cub::alignment_of<T>::value;
```
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:73]:
```suggestion
      constexpr ::cuda::std::size_t alignment = ::cuda::std::alignment_of<T>::value;
```

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    #else 
+    template<class K, class... Args>
+    cudaError_t __device__
+    doit_device(K, Args const&... ) const
+    {
+      return cudaErrorNotSupported;
+    }
+    #endif
+
+    __thrust_exec_check_disable__ // what is
```
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:140]:
```suggestion
    _LIBCUDACXX_DISABLE_EXEC_CHECK
```

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    template<class T>
+    size_t __device__
+    align_up(size_t offset) const
+    {
+      size_t alignment = THRUST_NS_QUALIFIER::cuda_cub::alignment_of<T>::value;
+      return alignment * ((offset + (alignment - 1))/ alignment);
+    }
+
+    size_t __device__ argument_pack_size(size_t size) const { return size; }
+    template <class Arg, class... Args>
```
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:78]:
```suggestion
    size_t __device__ argument_pack_size(size_t size) const { return size; }
    
    template <class Arg, class... Args>
```

File: cub/cub/detail/triple_chevron_launch.cuh (hunk 4 above)
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:109]:
```suggestion
    #ifdef CUB_RDC_ENABLED
```

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    cudaError_t __device__
+    doit_device(K, Args const&... ) const
+    {
+      return cudaErrorNotSupported;
+    }
+    #endif
+
+    __thrust_exec_check_disable__ // what is
+    template <class K, class... Args>
+    __host__ __device__ __forceinline__ // use this over THRUST_FUNCTION macro, seems to be convention in CUB
```
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:142]:
```suggestion
    __host__ __device__ __forceinline__
```

File: cub/cub/device/dispatch/dispatch_batch_memcpy.cuh [hunk 11]
```diff
@@ -48,7 +48,7 @@ _CCCL_IMPLICIT_SYSTEM_HEADER
 #include <cub/util_debug.cuh>
 #include <cub/util_ptx.cuh>
 
-#include <thrust/system/cuda/detail/core/triple_chevron_launch.h>
+ #include <cub/detail/triple_chevron_launch.cuh>
```
Comment (senior-zero) [inline cub/cub/device/dispatch/dispatch_batch_memcpy.cuh:51]: Here and in the remaining places:
```suggestion
#include <cub/detail/triple_chevron_launch.cuh>
```

File: cub/cub/detail/triple_chevron_launch.cuh [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+namespace detail
+{
+
+  struct _LIBCUDACXX_HIDDEN triple_chevron
+  {
+    typedef size_t Size;
+    dim3 const grid;
+    dim3 const block;
+    Size const shared_mem;
```
Comment (senior-zero) [inline cub/cub/detail/triple_chevron_launch.cuh:48]: important: this struct is still present in thrust headers. Please, replace it with `  using triple_chevron = cub::detail::triple_chevron;` in `thrust/thrust/system/cuda/detail/core/triple_chevron_launch.h`.

File: cub/cub/device/device_copy.cuh [hunk 13]
```diff
@@ -42,8 +42,7 @@ _CCCL_IMPLICIT_SYSTEM_HEADER
 
 #include <cub/device/dispatch/dispatch_batch_memcpy.cuh>
 
-#include <thrust/system/cuda/detail/core/triple_chevron_launch.h>
-
+#include <cub/detail/triple_chevron_launch.cuh>
```
Comment (senior-zero) [inline cub/cub/device/device_copy.cuh:45]: suggestion: I don't think this include is needed by this header. Please, remove it.

File: cub/cub/detail/triple_chevron_launch.cuh (hunk 8 above)
Comment (senior-zero) [reply cub/cub/detail/triple_chevron_launch.cuh:140]: @miscco I think we'll need _CCCL variant of this macro.

**Review by senior-zero — CHANGES_REQUESTED** [review]
Thank you for your contribution! There are a few minor comments below. Apart from that, not all of the https://github.com/NVIDIA/cccl/issues/568 requests are addressed. First of all, Thrust implementation is still present (instead of relying on alias). Secondly, there's no test of triple chevron.

File: cub/cub/detail/triple_chevron_launch.cuh (hunk 2 above)
Comment (senior-zero) [reply cub/cub/detail/triple_chevron_launch.cuh:51]: @ZelboK you are right, `CUB_RUNTIME_FUNCTION` should be used here.

File: cub/test/catch2_test_triple_chevron.cu [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+                  d_out,
+                  num_items,
+                  device_invoke);
+  }
+};
+
+struct cdp_invocable { 
+
+  template <class T>
+  CUB_RUNTIME_FUNCTION cudaError_t operator()(uint8_t* temp, size_t bytes, T d_in, T d_out, int n, bool on_device) const {
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:86]: important: your test is failing because chevron invoker modifies local variable `bytes` instead of a reference. This leads to the cdp helper allocating 0 bytes of temporary storage, so `temp` remains `nullptr` and the algorithm thinks it's still at the stage of querying temporary storage size.

File: cub/test/catch2_test_triple_chevron.cu [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    {
+        d_out[i] = d_in[i] * T{2};
+        if(blockIdx.x == 0 && threadIdx.x == 0) {
+            printf("After computation: d_out[%d] = %d\n", i, d_out[i]);
+        }
+    }
+}
+
+
+struct cdp_chevron_invoker {
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:32]: suggestion: I don't think this struct is required. You should be able to put everything from the `invoke` function inside the `cdp_invocable::operator()`.

File: cub/test/catch2_test_triple_chevron.cu [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+CUB_TEST("Triple Chevron with missing configuration returns cudaErrorMissingConfiguration", "[test][utils]") {
+  int n = 42;
+  thrust::device_vector<int> in(n, 21);
+  thrust::device_vector<int> out(n);
+  int *d_in  = thrust::raw_pointer_cast(in.data());
+  int *d_out = thrust::raw_pointer_cast(out.data());
+
+  auto chevron = cub::detail::triple_chevron(0, 0);
+  auto err = chevron.doit(mult_two_kernel<int>, d_in, d_out, n);
+  REQUIRE( CubDebug(cudaErrorMissingConfiguration) == cudaErrorMissingConfiguration );
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:180]: important:
```suggestion
  REQUIRE( err == cudaErrorInvalidConfiguration );
```

File: cub/test/catch2_test_triple_chevron.cu [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+  {
+    const auto actual   = static_cast<std::size_t>(thrust::count(out.begin(), out.end(), 10));
+    const auto expected   = static_cast<std::size_t>(n);
+    REQUIRE(actual==expected);
+  }
+  cudaEventDestroy(event);
+
+}
+
+CUB_TEST("Triple Chevron with missing configuration returns cudaErrorMissingConfiguration", "[test][utils]") {
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:171]: suggestion: the functionality we wanted to test is if the error is returned, so I'd rephrase that:
```suggestion
CUB_TEST("Triple chevron returns kernel launch failures", "[test][utils]") {
```

File: cub/test/catch2_test_triple_chevron.cu [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+
+  cudaDeviceProp deviceProperties;
+  cudaGetDeviceProperties(&deviceProperties, 0);
+  auto cap = deviceProperties.sharedMemPerBlock;
+
+  const int block_size = 2566666;
+  const int grid_size = (n * block_size - 1) / block_size;
+  auto chevron = cub::detail::triple_chevron(grid_size, block_size, cap); 
+  auto err = chevron.doit(mult_two_kernel<int>, d_in, d_out, n);
+  REQUIRE( CubDebug(cudaErrorInvalidConfiguration) == cudaErrorInvalidConfiguration );
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:198]: important:
```suggestion
  REQUIRE( err == cudaSuccess );
```

File: cub/test/catch2_test_triple_chevron.cu [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+  thrust::device_vector<int> in(n, 21);
+  thrust::device_vector<int> out(n);
+  int *d_in  = thrust::raw_pointer_cast(in.data());
+  int *d_out = thrust::raw_pointer_cast(out.data());
+
+  cudaDeviceProp deviceProperties;
+  cudaGetDeviceProperties(&deviceProperties, 0);
+  auto cap = deviceProperties.sharedMemPerBlock;
+
+  const int block_size = 2566666;
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:194]: important: we already have a test of invalid configuration above, this one should check that the dynamic shared memory is allocated:
```suggestion
  const int block_size = 256;
```

File: cub/test/catch2_test_triple_chevron.cu [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+  const int grid_size = (n * block_size - 1) / block_size;
+  auto chevron = cub::detail::triple_chevron(grid_size, block_size, cap); 
+  auto err = chevron.doit(mult_two_kernel<int>, d_in, d_out, n);
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:197]: suggestion: this kernel doesn't check if the dynamic shared memory is available and of requested size. Please, add `extern __shared__ char smem[];` to a kernel that will be used for this test and validate if it's of a proper size.

File: cub/test/catch2_test_triple_chevron.cu [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+  REQUIRE( CubDebug(cudaErrorInvalidConfiguration) == cudaErrorInvalidConfiguration );
+}
+
+CUB_TEST("Triple Chevron properly forwards parameters", "[test][utils]") {
+  double result;
+  double *d_result;
+
+  cudaMalloc(&d_result, sizeof(double));
+  auto chev = cub::detail::triple_chevron(1, 1); 
+  chev.doit(add_kernel, 5, 3.5f, d_result);
```
Comment (gevtushenko) [inline cub/test/catch2_test_triple_chevron.cu:207]: suggestion: we have different logic for hots and device launch. I'd suggest using cdp helper here as well.

**Comment by jrhemstad** [comment]
Closing this as stale. @ZelboK feel free to reopen with merge conflicts resolved if this is something that is still of interest to you.
